include/scale/detail/fixed_width_integer.hpp

```diff
@@ -6,37 +6,132 @@
 
 #pragma once
 
-#include <algorithm>
-#include <array>
-#include <cstdint>
-#include <vector>
+#include <type_traits>
 
-#include <boost/endian/arithmetic.hpp>
+#include <boost/endian/buffers.hpp>
+#include <boost/multiprecision/cpp_int.hpp>
 
 #include <scale/outcome/outcome_throw.hpp>
 #include <scale/scale_error.hpp>
-#include <scale/unreachable.hpp>
 
-namespace scale::detail {
+namespace scale {
+  class ScaleEncoderStream;
+  class ScaleDecoderStream;
+}  // namespace scale
+
+namespace scale {
+
+  using uint128_t = boost::multiprecision::checked_uint128_t;
+  using uint256_t = boost::multiprecision::checked_uint256_t;
+  using uint512_t = boost::multiprecision::checked_uint512_t;
+  using uint1024_t = boost::multiprecision::checked_uint1024_t;
+
+  template <typename T>
+  concept BigFixedWidthInteger =
+      std::is_same_v<std::remove_cvref_t<T>, uint128_t>
+      or std::is_same_v<std::remove_cvref_t<T>, uint256_t>
+      or std::is_same_v<std::remove_cvref_t<T>, uint512_t>
+      or std::is_same_v<std::remove_cvref_t<T>, uint1024_t>;
+
+  template <typename T, typename = void>
+  struct FixedWidthIntegerTraits;
+
+  // Для целочисленных типов
+  template <std::integral T>
+  struct FixedWidthIntegerTraits<T> {
+    static constexpr size_t bytes = sizeof(T);
+    static constexpr size_t bits = bytes * 8;
+  };
+
+  template <typename T>
+  struct FixedWidthIntegerTraits<T, std::enable_if_t<std::is_class_v<T>>> {
+   private:
+    static constexpr size_t compute_bytes() {
+      T temp{};
+      return temp.backend().size() * sizeof(temp.backend().limbs()[0]);
+    }
+
+   public:
+    static constexpr size_t bytes = compute_bytes();
+    static constexpr size_t bits = bytes * 8;
+  };
+
+  namespace detail {
+    template <typename To, typename From>
+      requires std::is_convertible_v<From, To>
+    To convert_to(From value) {
+      return static_cast<To>(value);
+    }
+
+    template <typename To, typename From>
+      requires boost::multiprecision::is_number<From>::value
+    To convert_to(const From &value) {
+      try {
+        return value.template convert_to<To>();
+      } catch (const std::runtime_error &e) {
+        // scale::decode catches std::system_errors
```

Comment:
Probably better to declare an custom exception type for SCALE library?

Reply:
That would require wrapper function for `.value()` (which throws `std::system_error`) which is used inside decoders, to rethrow outcome as new type